File: UnionFind/unionFind.cpp
/*
Implementation of the UnionFind data structure described here:
https://en.wikipedia.org/wiki/Disjounsigned-set_data_structure

Compiled with g++ 5.4.0:
	g++ -std=c++11 -Ofast -Wall -o unionFind unionFind.cpp

@2017 Florin Tulba (florintulba@yahoo.com)
*/

#include <string>
#include <fstream>
#include <sstream>
#include <iostream>
#include <iomanip>
#include <iterator>
#include <vector>
#include <map>
#include <memory>
#include <cassert>

using namespace std;

/// Data specific for each item of the Union Find
struct UfItem {
	unsigned ancestor;	///< the most distant known ancestor
	unsigned rank = 0U;	///< initial depth of a subtree rooted on this item (some children might migrate closer to the root of this item)

	UfItem(unsigned id) : ancestor(id) {}
};

/// The UnionFind class. It's the one in charge of validating the data input
class UF {
protected:
	vector<UfItem> items;	///< the managed items
	unsigned itemCount;		///< size of items
	unsigned groups;		///< current count of the groups formed from the items

	/// Checks if the provided index might be a valid element index
	bool validateIndex(unsigned idx) const {
		if(idx < itemCount)
			return true;

		cerr<<"Invalid element index: "<<idx<<endl;
		return false;
	}

	/// Find parent operation
	unsigned parentOf(unsigned id) {
		assert(validateIndex(id));

		unsigned *parentId = nullptr;
		while(*(parentId = &items[id].ancestor) != id)
			id = *parentId = items[*parentId].ancestor;

		return id;
	}

public:
	/// Create itemCount items that are initially separated
	UF(unsigned itemCount_) : itemCount(itemCount_), groups(itemCount_) {
		items.reserve(itemCount);
		for(unsigned i = 0U; i<itemCount; ++i)
			items.emplace_back(i);

		cout<<" Initially: "<<*this<<endl;
		if(itemCount < 2U)
			cout<<"Note that this problem makes sense only for at least 2 elements!"<<endl;
	}

	/// Are items id1 and id2 connected ?
	bool connected(unsigned id1, unsigned id2) {
		if( ! validateIndex(id1) || ! validateIndex(id2))
			return false;

		return parentOf(id1) == parentOf(id2);
	}

	/// Connect id1 & id2
	void join(unsigned id1, unsigned id2) {
		cout<<setw(3)<<id1<<" - "<<setw(3)<<id2<<" : ";
		if( ! validateIndex(id1) || ! validateIndex(id2))
			return;

		id1 = parentOf(id1); id2 = parentOf(id2);
		if(id1 == id2) {
			cout<<*this<<endl;
			return;
		}

		const unsigned rank1 = items[id1].rank, rank2 = items[id2].rank;
		if(rank1 < rank2)
			items[id1].ancestor = id2;
		else
			items[id2].ancestor = id1;

		if(rank1 == rank2)
			++items[id1].rank;

		--groups;

		cout<<*this<<endl;
		if(groups == 1U)
			cout<<"All elements are now connected!"<<endl;
	}

	friend ostream& operator<<(ostream &os, UF &uf) {
		os<<setw(3)<<uf.groups<<" groups: ";
		map<unsigned, vector<unsigned>> mapping;
		for(unsigned i = 0U, lim = uf.itemCount; i<lim; ++i)
			mapping[uf.parentOf(i)].push_back(i);
		for(const auto &group: mapping) {
			os<<group.first<<'{';
			copy(begin(group.second), end(group.second), ostream_iterator<unsigned>(os, " "));
			os<<"\b}  ";
		}
		return os;
	}
};

/// Provides the problem input data
class ScenarioProvider {
protected:
	ifstream ifs;			///< stream providing the test scenario
	unsigned items = 0U;	///< number of elements from the described scenario
	bool valid = true;		///< becomes false as soon as IO / parsing errors are found in the scenario file

public:
	typedef pair<unsigned, unsigned> ElemPair;	///< the indices from a pair of elements

	/// Reads itemsCount from 'testScenario.txt'
	ScenarioProvider() {
		static const string scenarioFile("testScenario.txt");
		ifs.open(scenarioFile);
		if( ! ifs) {
			cerr<<"Couldn't open the scenario file: "<<scenarioFile<<endl;
			valid = false;
			return;
		}
		string line;
		while(getline(ifs, line)) {
			if((line.empty() || (line.size() == 1ULL && line[0] == '\r')) || line[0] == '#')
				continue;	// Ignore empty lines or lines containing comments (these start with '#')

			istringstream iss(line);
			if( ! (iss>>items)) {
				cerr<<"Couldn't read the items count from the scenario file! Please correct this error and then try again!"<<endl;
				valid = false;
			}

			return;
		}
	}

	inline unsigned itemsCount() const { return items; }

	/// Reads next pair of element indices
	unique_ptr<ElemPair> nextElemPair() {
		if( ! valid) {
			cout<<"Please correct previous errors before calling ScenarioProvider::nextElemPair()!"<<endl;
			return nullptr;
		}

		if(ifs.eof()) {
			cout<<"No other pairs to connect in this scenario!"<<endl;
			return nullptr;
		}

		string line;
		while(getline(ifs, line)) {
			if((line.empty() || (line.size() == 1ULL && line[0] == '\r')) || line[0] == '#')
				continue;	// Ignore empty lines or lines containing comments (these start with '#')

			istringstream iss(line);
			unsigned idx1, idx2;
			if( ! (iss>>idx1>>idx2)) {
				cerr<<"Couldn't read the next pair of element indices from the scenario file! Please correct this error and then try again!"<<endl;
				valid = false;
				break;
			}

			//return make_unique<ElemPair>(idx1, idx2); // C++14
			return unique_ptr<ElemPair>(new ElemPair(idx1, idx2));
		}

		return nullptr;
	}
};

int main() {
	ScenarioProvider sp;
	UF uf(sp.itemsCount());
	unique_ptr<ScenarioProvider::ElemPair> elemPair;
	while(nullptr != (elemPair = sp.nextElemPair()))
		uf.join(elemPair->first, elemPair->second);

	return 0;
}

File: UnionFind/unionFind.R
#
# Implementation of the UnionFind data structure described here:
# 		https://en.wikipedia.org/wiki/Disjoint-set_data_structure
#
# @2017 Florin Tulba (florintulba@yahoo.com)
#


# Find parent operation
parentOf <- function(uf, id) {
	ufTemp <- uf
	parentId <- ufTemp[id, 'Ancestor']
	while(id != parentId) {
		ufTemp[id, 'Ancestor'] <- ufTemp[parentId, 'Ancestor']
		id <- ufTemp[id, 'Ancestor']
		parentId <- ufTemp[id, 'Ancestor']
	}
	eval.parent(substitute(uf <- ufTemp))
	as.integer(id)
}

# Connect id1 & id2
join <- function(uf, id1, id2) {
    # the decrements ensure same output as in programming languages with 0-based indexing
    cat(sprintf('%3d - %3d :', id1-1L, id2-1L))
	ufTemp <- uf
	id1 <- parentOf(ufTemp, id1)
	id2 <- parentOf(ufTemp, id2)
	if(id1 == id2) {
		eval.parent(substitute(uf <- ufTemp))
		return
	}

	rank1 <- ufTemp[id1, 'Rank']
	rank2 <- ufTemp[id2, 'Rank']
	if(rank1 < rank2)
		ufTemp[id1, 'Ancestor'] <- id2
	else
		ufTemp[id2, 'Ancestor'] <- id1

	if(rank1 == rank2)
		ufTemp[id1, 'Rank'] <- ufTemp[id1, 'Rank'] + as.integer(1)

	eval.parent(substitute(uf <- ufTemp))
}

# Displays the UnionFind content without mutating it (mutator method 'parentOf' will produce a local copy of uf)
strUf <- function(uf) {
	mapping <- sapply(ids, function(id) c(ParentId=parentOf(uf, id), Id=id))
	mapping <- mapping[,order(mapping['ParentId',])]
	cat(sprintf("%3d", length(unique(mapping[1,]))), 'groups:')
	i <- 1; parentId <- mapping['ParentId', i]
	repeat {
	    # the decrements ensure same output as in programming languages with 0-based indexing
		cat(parentId-1L, '{', mapping['Id', i]-1L, '')
		repeat {
			i <- i + 1
			if(i > n)
				break
			newParentId = mapping['ParentId', i]
			if(parentId != newParentId) {
				parentId <- newParentId
				break
			}
			# the decrements ensure same output as in programming languages with 0-based indexing
			cat(mapping['Id', i]-1L, '')
		}
		cat('}')
		if(i > n)
			break
		cat(', ')
	}
}

conn <- file('testScenario.txt', 'r')

source('../common/util.R') # for loading 'nextRelevantLine'
if(length(line <- nextRelevantLine(conn)) > 0) {
    n <- as.integer(line)			# items count
    if(n < 0L) {
        cat("Items count should be >= 0!\n")
    } else {
        if(n < 2L)
            cat('Note that this problem makes sense only for at least 2 elements!\n')

        ids <- 1L:n 					# items' ids
        ancestors <- ids				# initial items are also their own root ancestors

        # The rank of an item is the depth of a subtree rooted on this item (some children might migrate closer to the root of this item)
        ranks <- rep(0L, n)	# initial items' ranks are all 0

        # data frame with columns ancestor, rank
        uf <- data.frame(Ancestor=ancestors, Rank=ranks)
        cat(' Initially:'); cat(strUf(uf), '\n')

        noIssues <- TRUE
        while(length(line <- nextRelevantLine(conn)) > 0) {
            tokens <- strsplit(line, ' ', fixed = TRUE)[[1]]
            if(length(tokens) != 2) {
                cat(sprintf('Line "%s" contains less/more than 2 items!\n', line))
                noIssues <- FALSE
                break
            }

            idx1 <- as.integer(tokens[1])
            idx2 <- as.integer(tokens[2])
            if(idx1 < 0 || idx2 < 0 || idx1 >= n || idx2 >= n) {
                cat(sprintf('Line "%s" contains invalid indices!\n', line))
                noIssues <- FALSE
                break
            }

            # the increments ensure 1-based indexing as required in R
            join(uf, idx1+1L, idx2+1L); cat(strUf(uf), '\n')
        }
        if(noIssues)
            cat('No other pairs to connect in this scenario!')
    }
} else {
    cat("Couldn't read the items count!\n")
}

close(conn)
